### PR TITLE
CHE-5174 Improving search results 

### PR DESCRIPTION
**Pull request by vparfonov**
### What does this PR do?
Improving search results:
- line numbers are shown for matches
- multiple matches in the same file are shown individually
- when you click on a search result, the editor loads the file AND jumps straight to the matching line with all instances of the search term already highlighted

che-5174

### What issues does this PR fix or reference?
#5174 

related to https://github.com/eclipse/che-dependencies/pull/50

#### Changelog
ProjectService#search() now return different objects, now it will be `List<SearchResultDto>` instead of `List<ItemReference>`

#### Release Notes
Improving search results:
- line numbers are shown for matches
- multiple matches in the same file are shown individually
- when you click on a search result, the editor loads the file AND jumps straight to the matching line with all instances of the search term already highlighted

NOTE: ProjectService#search() now return different objects, now it will be `List<SearchResultDto>` instead of `List<ItemReference>`

### CODE REVIEW

**Review by RomanNikitenko — APPROVED** [review]

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/resources/SearchResult.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.eclipse.che.api.project.shared.Constants;
+import org.eclipse.che.api.project.shared.SearchOccurrence;
+import org.eclipse.che.api.project.shared.dto.SearchOccurrenceDto;
+import org.eclipse.che.api.project.shared.dto.SearchResultDto;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Vitalii Parfonov
```
Comment (skabashnyuk) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/resources/SearchResult.java:22]: javadoc

**Review by evidolob — APPROVED** [review]

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/project/ProjectServiceClientImpl.java [hunk 2]
```diff
@@ -164,7 +162,13 @@ protected ProjectServiceClientImpl(WsAgentStateController wsAgentStateController
         return reqFactory.createGetRequest(url + queryParameters.toString().replaceFirst("&", "?"))
                          .header(ACCEPT, MimeType.APPLICATION_JSON)
                          .loader(loaderFactory.newLoader("Searching..."))
-                         .send(unmarshaller.newListUnmarshaller(ItemReference.class));
+                         .send(unmarshaller.newListUnmarshaller(SearchResultDto.class)).then(
+                        (Function<List<SearchResultDto>, List<SearchResult>>)arg -> {
```
Comment (tolusha) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/project/ProjectServiceClientImpl.java]: I prefer not to have variable named `arg`. It is confused.

File: wsagent/che-core-api-project/pom.xml [hunk 3]
```diff
@@ -80,6 +80,11 @@
         </dependency>
         <dependency>
             <groupId>org.apache.lucene</groupId>
+            <artifactId>lucene-highlighter</artifactId>
+            <version>5.2.1</version>
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/pom.xml]: remove version

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/util/ReadFileUtils.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.LineNumberReader;
+import java.io.RandomAccessFile;
+
+import static java.lang.Character.toChars;
+import static java.lang.System.lineSeparator;
+
+/**
+ * @author Vitalii Parfonov
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/util/ReadFileUtils.java]: javdoc? May be to put this somewhere in Commons?

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/search/presentation/FoundOccurrenceNode.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+            updatePresentation(nodePresentation);
+            return nodePresentation;
+        }
+        if (update) {
+            updatePresentation(nodePresentation);
+        }
+        return nodePresentation;
+    }
+
+    /** {@inheritDoc} */
```
Comment (tolusha) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/search/presentation/FoundOccurrenceNode.java:121]: Redundant.

File: wsagent/che-core-api-project/src/test/java/org/eclipse/che/api/vfs/search/impl/ReadFileUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Test(expectedExceptions = IOException.class,
+          expectedExceptionsMessageRegExp = "File is not long enough")
+    public void genLineByOffsetShouldFail() throws Exception {
+        final Path tempFile = Files.createTempFile("my", "_test");
+        Files.write(tempFile, TEST_CONTENT2.getBytes());
+        ReadFileUtils.getLine(tempFile.toFile(), 10000);
+
+    }
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/test/java/org/eclipse/che/api/vfs/search/impl/ReadFileUtilsTest.java]: cleanup?

File: wsagent/che-core-api-project/pom.xml (hunk 3 above)
Comment (tolusha) [inline wsagent/che-core-api-project/pom.xml]: Specific version? CQ?

Comment (vparfonov) [reply wsagent/che-core-api-project/pom.xml]: fixed

Comment (vparfonov) [reply wsagent/che-core-api-project/pom.xml]: fixed

File: wsagent/che-core-api-project/src/test/java/org/eclipse/che/api/vfs/search/impl/ReadFileUtilsTest.java (hunk 6 above)
Comment (vparfonov) [reply wsagent/che-core-api-project/src/test/java/org/eclipse/che/api/vfs/search/impl/ReadFileUtilsTest.java]: fixed

**Review by tolusha — APPROVED** [review]

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/project/ProjectServiceClientImpl.java (hunk 2 above)
Comment (vparfonov) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/project/ProjectServiceClientImpl.java]: fixed

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**